design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+
+```
+renku dataset stage --copy very-large-data
+```
+
+The default location is to be `.renku/mounts/<name>` but can be specified with
+the `--location` flag in case it is preferred to take the data to an external
+drive or another filesystem.
+
+Once it's staged, it can be found locally and used in workflows:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      url: s3://server/path/to/<bucket>
+      location: .renku/mounts/<bucket>
+```
+
+The data is accessible under `data/very-large-data/<bucket>` in the project's directory.
+
+If the data is already mounted by some other means, `dataset stage` can be used
+to let Renku know about it:
+
+```
+renku dataset stage --existing s3://server/path/to/<bucket>:/mounts/bucket
+```
+
+#### Using external data _without_ Renku datasets
+
+We should also consider the case where creating a dataset is too much effort or
+does not fit the stage of the project. In such cases, Renku should allow the
+user to bring in and reference external data. For this, we should review the
+existing `storage` subcommand. For example, to bring in data from an S3 bucket,
+the user would do
+
+```
+renku storage add s3://server/path/to/<bucket>
+-- output --
+<bucket> added as external storage
+```
+
+This would not immediately do anything with the data, but only create the
+`DataStorage` metadata object. For bringing in the data, the user would need
+to do
+
+```
+renku storage stage --copy <bucket> path/to/bucket
+```
+
+The filesystem then looks like
+
+```
+.renku/mounts/<bucket>
+path/to/bucket --> .renku/mounts/<bucket>
+```
+
+The `path/to/bucket` is in `.gitignore`.
+
+The above can be condensed in
+
+```
+renku storage add --copy --to path/to/bucket s3://server/path/to/<bucket>
+```
+
+This is stored in the metadata so when the project is cloned elsewhere, the
+relative location of the data is fixed. So, e.g.
+
+```
+renku storage ls
+-- output --
+Type |             URL              |        Location        |  Project path
+-------------------------------------------------------------------------------
+S3   | s3://server/path/to/<bucket> | .renku/mounts/<bucket> | path/to/<bucket>
+```
+
+To quickly get access to all of the external data when cloning the project
+in a new location, the user can easily mount/copy all the known remote data sources:
+
+```
+renku storage stage --all --copy
+```
+
+Note that in the case of remote storage (as opposed to datasets) there is no
+contradiction in writing directly to the directory where the data source is made
+available in the local project. If the data source is mounted, then the changes
+are reflected on the remote location immediately; if the data is copied, it
+should be possible to implement a `sync` command that would copy upstream new or
```

Comment:
What happens when we have conflicts during this sync operation? I.e. when a remote file is newer than a local file? Also if one person has the same dataset mounted with fuse (i.e. without copying) do we allow another person to add the data with either the `--copy` flag or without.

Reply:
A few things that are somewhat related are:
1. Handling buckets to which the user has no write access. In this case we should make sure that after mounting the files (with or without `--copy`) the user is also not allowed to write to them
2. We should have a flag in the `storage`/`dataset` `add` command to indicate whether a dataset is read-only or read/write. And this should be compatible with the access level that the credentials provide. 
3. There may be tricky corner cases with granular access control policies that I think in the case of S3 can apply only to specific files or subfolders.